### ISSUE TITLE
reproducibly capture the iterative development of an analysis

### ISSUE THREAD

**Issue by rgayler** [ropensci/ozunconf17#10]
This may be somewhat related to @MilesMcBain issue on [provenance tooling](https://github.com/ropensci/unconf17/issues/23)

The standard use case for reproducible research tools (Rmarkdown and friends) seems to be the final reporting of some analysis. What gets reported in papers is typically a minimal, apparently logical, and direct progression through the steps needed to get the results to convince the reader.

However, the process of arriving at that analysis and results may have been very convoluted and iterative. In scientific analytics there are typically many choices to be made along the way which may have been supported by their own unreported analyses. in the spirit of reproducibility it would be helpful if these decisions and the reasoning behind these decisions could be captured. I presume that this would typically be captured in the laboratory notebooks of wet-lab scientists.

I face a similar issue in my day job. I tend to work in a heavily regulated environment where there are multiple layers of validation/audit/regulation. It is not enough to build a model - I have to also document what choices I made, and why I made the choices I did.

In the type of projects I do, most of the work goes into prepping the data for the actual analysis. I can spend weeks incrementally changing the data prep process as I discover more wrinkles in the data. Sometimes these wrinkles are detected by the actual data analysis, but more often they are detected by the copious data prep analyses I do to understand the data.

The (final) actual analysis only depends on the data produced by the final version of the prepping code (which reflects all the stuff I have learned incrementally about the data). For complete reproducibility, and also to capture what I learned about the data, I feel I ought to be able to reproduce all the incremental data prepping analyses I did along the way. But each data prep analysis depends on the (partly-flawed) data prep process that existed at the time. So for complete reproducibility, I would need to keep multiple versions of the data prep code, the data produced by each version of the data prep code, and the prep analyses that were run against that data to conclude that the data prep process needed improvement.

So far, I have tended to keep each data prep analysis as a separate markdown notebook, with the data prep code cut and pasted from document to document and updated. I am far from convinced this is the best way to do it. I am up for discussions of how to organise/capture this kind of iterative development process.

**Comment by adam-gruer** [comment]
I think git could have a role to play here - given that each commit is a snapshot of a project at a point in time.

**Comment by rgayler** [comment]
Thanks, Adam. I don't know enough about git to be sure, but I think my
problem is that I don't want to track the history of a single artifact, but
want multiple interrelated artifacts that each have their own development
history that might usefully be tracked with git. I need to draw a diagram
to show what I think I want - words aren't adequate. I will do that when I
have some time.

I am prepared to believe:
* that what I want is trivially easy with git - i just didn't know how to
do it, or
* what I want is dumb and there is a much more straight forward way to get
the same effect.

I am looking forward to some discussions once i have a sensible diagram to
talk about.

Ross

**Comment by timchurches** [comment]
I think git can do everything you need. You can choose to bundle arbitrary ensembles of files (artefacts) into a commit (not necessarily all changed files) and describe in the commit message (and longer commit note if required) what you did with respect to all those artefacts, as a gestalt. That way you can see all the changes to each file in the context of changes to related files. If you adopt a naming scheme embedded in your commit messages, eg prepend “[A]”, “[B]” etc each relating to an aspect of your analysis, you can easily filter all the commits to see only those that relate to a particular aspect.

**Comment by MilesMcBain** [comment]
Your workflow is all too familiar @rgayler. 

Just recently I've been really impressed by @jdblischak's [workflowr](https://github.com/jdblischak/workflowr) which has some built in provenance support. Check out [my test drive of it here](https://milesmcbain.github.io/keras_facial_keypoints/baseline_model.html).

It does two useful things re provenance: Embeds the Github commit id up the top, and embeds the session info at the bottom.

The downside is it is geared toward a very specific workflow that will either work for you or not - and this tends to be the main issue I see with development in this area: Developing tools that make data and code provenance tracking easy while being flexible enough about the workflow to be used.

It doesn't address data versioning and this becomes a problem with larger datasets that cannot be pulled from a repository every time the analysis is to be run. A while ago I put some thoughts down about this [here](https://github.com/MilesMcBain/journalr/blob/master/Journalling_tool_proposal.Rmd)

The solution I conceived of boils down to providing a way to capture analysis/data munging metadata to be stored alongside the analysis code. My dream would be able to be to query that metadata to explore the decisions made and the impact they had. I'm not talking about searching comments here. I thinking queries like: "*At what point was column X introduced? What was the impact on model metrics?*", *"How many variations of column Y exist in the analysis history? How do the summaries of those versions differ?*"

@rgayler Maybe you have some thoughts on metadata that could be captured to document the decisions you make?

**Comment by jdblischak** [comment]
> Just recently I've been really impressed by @jdblischak's workflowr which has some built in provenance support. Check out my test drive of it here.

> It does two useful things re provenance: Embeds the Github commit id up the top, and embeds the session info at the bottom.

Thanks, @MilesMcBain! I'm glad you found workflowr interesting. You are correct that the goal of workflowr is to help with @rgayler's description of iteratively updating a data analysis project.

> The downside is it is geared toward a very specific workflow that will either work for you or not - and this tends to be the main issue I see with development in this area: Developing tools that make data and code provenance tracking easy while being flexible enough about the workflow to be used.

Agreed that this is the essential trade-off. I abstract away the details of running Git commands and building a static website so that the user can focus solely on writing R code to analyze their data, but this requires using a specific directory structure so that the workflowr functions can guess what needs to be done automatically.

workflowr is a bit more flexible than the documentation describes. The directory `analysis/` can be renamed to anything, same with the `docs/` directory (the user would need to manually run `git mv`, which is why I don't yet advertise this feature more prominently). To get technical, workflowr searches for an `.Rproj` file, and then searches the top-level directories for a `_site.yml` file.

> It doesn't address data versioning and this becomes a problem with larger datasets that cannot be pulled from a repository every time the analysis is to be run. A while ago I put some thoughts down about this here

Right, there is still a lot of responsibility left to the user regarding the data (in order to be flexible). If the user doesn't commit the data files, then the commit ID inserted at the top of the HTML file is meaningless.

For larger files, I am using [git lfs](https://git-lfs.github.com/) for one of my workflowr projects. The main downsides are 1) additional setup steps to configure git lfs and 2) costs money on GitHub.

> The solution I conceived of boils down to providing a way to capture analysis/data munging metadata to be stored alongside the analysis code. My dream would be able to be to query that metadata to explore the decisions made and the impact they had. I'm not talking about searching comments here. I thinking queries like: "At what point was column X introduced? What was the impact on model metrics?", "How many variations of column Y exist in the analysis history? How do the summaries of those versions differ?"

Very cool! I'll be interested to see what you develop.

**Comment by cboettig** [comment]
Great thread and interested to see where this goes!

My short take-away is that using git does a reasonable job of letting you return to a particular point in your project's history; ideally if you are also documenting the decisions in a notebook format as @rgayler discusses, since these aren't always reflected in code & commit messages.  (though GitHub's issue tracker kinda operates like this).  I used a separate notebook repo for this purpose for a long time, but have found it simpler (particularly as I now work most often on multiple projects with different people) to just keep a `notebook` directory in each individual git repo.  (More on that here: http://www.carlboettiger.info/2015/12/31/reflections-on-the-open-lab-notebook.html)

Pain points / weak links in the "just use a git repo" approach for me are: large data files; as others have discussed, and 'external' dependencies, particularly dependencies to some other project I have in a separate repo.  (Careful planning or simple duplication can help get around the latter).

Regarding the data workflow & data versioning, I'd recommend looking at https://github.com/ropenscilabs/datastorr if you haven't yet.  Managing data in a separate data repo has some upsides (keeps code repo smaller) and downsides (coordinating repos), but the github "release" model with GitHub, coupled with Zenodo archiving for versioned DOIs is really compelling to me.

**Comment by rgayler** [comment]
Jumping back into this - the day job keeps getting in the way.

I really need to draw a diagram of a typical project - maybe this weekend. 

A typical project for me (which may be idiosyncratic and irrelevant to everyone else) tends to consist of multiple related subprojects. I build a data preparation pipeline that might ultimately consist of ten chained blocks. This data preparation pipeline is pretty much pure code. There are comments, but no extended explanatory text in the code. I build this pipeline up sequentially - one block at a time. When I am working on a block it is the usual programming iterative revision stuff. Occasionally I may need to go back and modify an earlier block. It is only the final state of each of the pipeline blocks that is relevant. I have never had any need to go back to an earlier state on these.

While each pipeline block is being developed I simultaneously have an associated analysis subproject. This subproject takes the data output from the pipeline block and does multiple analyses (each possibly with some more local data prep). The purpose of this subproject is to test the pipeline block, to make sure it is doing what I think it should be and to gain insights into the (usually horrible) data to inform the construction of the next pipeline block. The current pipeline block and associated analyses are jointly updated. The analyses are done in R notebooks and include lots of text which capture my thoughts and conclusions about the data with respect to how it is prepared. Once again, it is only the final state of the notebook which is relevant. I have never had a real need to return to earlier versions of the notebook. The findings in the notebooks document the pipeline design decisions - "The pipeline does X because the data looks like Y. I could have done P but rejected it because of Q in the data". This kind of documentation is probably too long form for a commit message.

Occasionally, when I get to the analyses accompanying a later stage of the pipeline I may discover that I need to change something in an earlier block. So i go back to the earlier block, change it, then re-run all subsequent pipeline blocks and associated analyses to check that none of my conclusions have changed as a consequence.

So my projects tend to be accumulations of components rather than the modification of components over time. (Yes, individual components have an internal temporal evolution but I am only interested in the final state of each and haven't felt hampered by not using version control. Where VC might be useful would be if I needed to modify the whole project in the future because of changes to the data source.) Although the components tend to be sequentially developed over time, the final state of interest has them all present simultaneously because they are components of the complete solution. Some of the components comprise the data preparation and modelling pipeline that is the primary output of the project. Many of the components are secondary data munging and analyses used to guide the design of the primary pipeline.

Coming back to the earlier responses:

@timchurches 

> I think git can do everything you need. ... That way you can see all the changes to each file in the context of changes to related files. ... you can easily filter all the commits to see only those that relate to a particular aspect.

I am not convinced git is *necessary* for my purposes. Only the final version of everything is of external interest. Nobody else will care about the path by which it reached that final state. git may help me not lose stuff as I am developing it, but once the project is complete I could throw away the git history and not lose anything that users care about. Also, there is only one analyst, so git doesn't provide any coordination advantage here.

@MilesMcBain 

> [workflowr] is geared toward a very specific workflow that will either work for you or not - and this tends to be the main issue I see with development in this area: Developing tools that make data and code provenance tracking easy while being flexible enough about the workflow to be used.

Yes - maybe my workflow is too idiosyncratic to be of any general interest. Also, my experience with formalised processes is that they require a degree of OCD on the part of the user. But the more OCD required, the less likely users will stick with it and the less likely it is to fit any specific use case.

> The solution I conceived of boils down to providing a way to capture analysis/data munging metadata to be stored alongside the analysis code. My dream would be able to be to query that metadata to explore the decisions made and the impact they had. I'm not talking about searching comments here. I thinking queries like: "At what point was column X introduced? What was the impact on model metrics?", "How many variations of column Y exist in the analysis history? How do the summaries of those versions differ?"

This is more micro-level than I had in mind. The decisions I make and document tend to be about only some of the data (not every single variable) and tend to be based on a combination of data facts and external knowledge - "The data sample says X and I have an external belief that X is not true - so I will not use the implicated variables in modeling". Also, metadata like "At what point was column X introduced?" may be useful in the context of an operational system that is modified over time, it's not so clear that it's useful in a one-off development context where the order in which things happened may have been pretty random.

@jdblischak 

Thanks - I will look at workflowr.

@cboettig 

> using git does a reasonable job of letting you return to a particular point in your project's history; 

Yes. Going back to my workflow, I see this as insurance against screw-ups in the development process but not being directly necessary to the final product. The end product documents the important design decisions made. There is a historical ordering of those decisions due to the development process, but that history is not relevant for use of the end-product. The end-product lays out all those decisions simultaneously in an ahistorical way.

> you are also documenting the decisions in a notebook format ... since these aren't always reflected in code & commit messages.

Yes - in my current workflow these tend to be a paragraph or two following on from some analyses, so I want them to be with the analytical results (like [this](https://github.com/ropensci/ozunconf17/issues/13)).

> I now ... keep a notebook directory in each individual git repo

In my current workflow I (mostly) have a separate notebook per data pipeline stage, analysing the data output from that stage, and informing the design of the next pipeline stage. While these notebooks mostly grow by extension over time, I don't explicitly capture the temporal aspect because they are intended to record the analyses and consequent design reasoning rather than the chronology of individual actions.

> More on that here: http://www.carlboettiger.info/2015/12/31/reflections-on-the-open-lab-notebook.html

Thanks - I'll go read it.

> Pain points / weak links in the "just use a git repo" approach for me are: ... 'external' dependencies, particularly dependencies to some other project I have in a separate repo.

I think part of the problem with using git my current workflow is how to treat it as a project. I should probably treat the whole thing as a single project, but it almost feels like the data preparation pipeline should be one project and each of the stage-specific-analyses-and-notebooks should be a separate project. As I said, I have actually managed to get by without git for this stuff without feeling I am missing out on too much.

**Comment by timchurches** [comment]
@rgayler 

> I am not convinced git is necessary for my purposes. Only the final version of everything is of external interest. Nobody else will care about the path by which it reached that final state. git may help me not lose stuff as I am developing it, but once the project is complete I could throw away the git history and not lose anything that users care about. Also, there is only one analyst, so git doesn't provide any coordination advantage here.

However, that seems to be at odds with the requirement that you originally stated:

@rgayler
>  The (final) actual analysis only depends on the data produced by the final version of the prepping code (which reflects all the stuff I have learned incrementally about the data). For complete reproducibility, and also to capture what I learned about the data, I feel I ought to be able to reproduce all the incremental data prepping analyses I did along the way. But each data prep analysis depends on the (partly-flawed) data prep process that existed at the time. So for complete reproducibility, I would need to keep multiple versions of the data prep code, the data produced by each version of the data prep code, and the prep analyses that were run against that data to conclude that the data prep process needed improvement.
 
I still think that you just need to bite the bullet and use git, and find a workflow with it that suits your needs. For example, your requirement "...so for complete reproducibility, I would need to keep multiple versions of the data prep code, the data produced by each version of the data prep code, and the prep analyses that were run against that data to conclude that the data prep process needed improvement..." could be met by using annotated tags in git (see https://git-scm.com/book/en/v2/Git-Basics-Tagging) to enable you to quickly re-establish your complete data cleaning and analysis environment as it was at any tagged point, complete with digitally signed checksums to prove that's how it was at that point.

**Comment by MilesMcBain** [comment]
@timchurches I could see branching providing a better solution than tagging. It matches the process that @rgayler describes in his first post: 'the garden of forking paths' we know so well.

But that is just my subjective opinion - and that's actually the main comment I want to make here: We can say "It's obvious! Use git!", but in no way was git designed for R analysts or scientists. We're shoehorning a closely related but different set of problems into git and there are many ways that could be done. 

I prefer to start from the other direction in this case: What idealistic solutions to this provenance problem can we conceive of? Git is then just a potential back end to be used in satisfying those solutions... hopefully buried under layers of wrapping. I see a strong connection between this issue and #16.

**Comment by goldingn** [comment]
@MilesMcBain I totally agree! I'd love to spend unconf time working out what abstractions and git workflows are useful to analysts and packaging them up nicely.

**Comment by goldingn** [comment]
Also, extending the idea in #16 to fully integrate data versioning and package/external dependency versioning would add a lot of value. There are packages that solve the technical problems, but putting them all behind a single interface would be so awesome.

**Comment by timchurches** [comment]
@MilesMcBain 
> I could see branching providing a better solution than tagging. It matches the process that @rgayler describes in his first post: 'the garden of forking paths' we know so well.

Yes, definitely use branches. My rule is that all new work is done on a branch. I was just suggesting tagging the branch merges as waypoints. It makes it so much easier to revert to a particular waypoint when a project has thousands of commits. 

I agree though that git is as much of a problem as it is a solution, at least at first, hence many people's reluctance to use it. Every time we have tried to teach git to postgrad students, it has been at least a partial fiasco, although some do get it and start to use it. The git GUI interface in RStudio helps, and in RStudio v1.1 it now supports branches, which means that many users can do just about everything they need through it.

The real need, as others have mentioned, is data versioning/coordination. In the health domain in which I work, this is complicated by the need to maintain confidentiality and thus keep data files secure, and that implies encryption, at least for transmission and any off-site storage, such as on GitHub (even in private repos). git LHS solves the large file storage problem ( https://git-lfs.github.com ), and can be used with a large number of back-end storage options (see https://github.com/git-lfs/git-lfs/wiki/Implementations ). Working out how to integrate this into R workflows could be a useful project.

**Comment by dicook** [comment]
A related topic is having a history of exploring data using interactive visualisation, so that it could be reproduced. 

I think a long time ago someone provided an .Rhistory editor, to be able to annotate, delete, ... with the goal to be document the exploration process. It would be really useful to see how many dead-ends one goes down and what leads to switching gears.

**Comment by rgayler** [comment]
I have finally had a chance to think a bit harder about my typical projects and what I really want. The write-up is [here](http://rpubs.com/r_gayler/workflow).
